Subtle preprocessor / linker problem with custom build configs

The project at

https://www.dropbox.com/s/ghyfybpieh1jb10/ReactiveBuildShenanigans.zip

builds against commit 3a16b8f0c179 of RAC (xcode6 branch) and exposes a nasty storm of Xcode build failovers, preprocessor assumptions, and macro patches of Apple API that culminates in a linker error.

Open the workspace and inspect the MasterProject scheme. Note the Archive step uses a custom BORK configuration. (RAC only has the standard Debug and Release configs.) Now select Product > Archive and you should see a linker error:

Undefined symbols for architecture armv7:
  "_rac_dispatch_async", referenced from:
      _main in main.o

Here's what's happening:

1) The selected BORK configuration does not exist in the RAC project, so RAC falls back to Release. (How Xcode chooses the fallback config is a mystery, and AFAICT there's no way to deterministically guide the decision)
2) As designed, the Release build of libReactiveCocoa lacks the rac_dispatch functions due to the #ifdef DEBUG in RACBacktrace.m
3) MasterProject builds with DEBUG=1, and because it imports RACBacktrace.h (via ReactiveCocoa.h), the preprocessor patches dispatch_*
4) The linker balks because the symbols the preprocessor said would be there aren't (see #2)

There are multiple xcodebuild shortcomings at play here, but switching to an RAC_DEBUG cflag would solve this issue entirely... unless someone defined an RAC_DEBUG in their project, which is much less likely, and less defensible, than the DEBUG=1 that comes with every new Xcode project.

You could also unconditionally include the rac_dispatch symbols, but if you had done that in the first place then a lot of people would inadvertently ship patched GCD functions without ever knowing it, and that would be bad. I only bring that up to ensure you don't do it :-)

[3a16b8f0c179 / xcode6] now has ReactiveCocoa.h importing RACBacktrace.h automatically, which is how we found the issue. But this will repro on any branch where the "parent" project happens to import RACBacktrace.h, either directly or indirectly.

--- a/ReactiveCocoaFramework/ReactiveCocoa/RACBacktrace.h
+++ b/ReactiveCocoaFramework/ReactiveCocoa/RACBacktrace.h
@@ -6,7 +6,7 @@
 //  Copyright (c) 2012 GitHub, Inc. All rights reserved.
 //
 
-#ifdef DEBUG
+#ifdef RAC_DEBUG_BACKTRACE
 
 extern void rac_dispatch_async(dispatch_queue_t queue, dispatch_block_t block);
 extern void rac_dispatch_barrier_async(dispatch_queue_t queue, dispatch_block_t block);

--- a/ReactiveCocoaFramework/ReactiveCocoa/RACBacktrace.m
+++ b/ReactiveCocoaFramework/ReactiveCocoa/RACBacktrace.m
@@ -12,7 +12,7 @@
 
 #define RAC_BACKTRACE_MAX_CALL_STACK_FRAMES 128
 
-#ifdef DEBUG
+#ifdef RAC_DEBUG_BACKTRACE
 
 // Undefine the macros that hide the real GCD functions.
 #undef dispatch_async

--- a/ReactiveCocoaFramework/ReactiveCocoaTests/RACBacktraceSpec.m
+++ b/ReactiveCocoaFramework/ReactiveCocoaTests/RACBacktraceSpec.m
@@ -14,7 +14,7 @@
 #import "RACSequence.h"
 #import "RACSignal+Operations.h"
 
-#ifdef DEBUG
+#ifdef RAC_DEBUG_BACKTRACE
 
 static RACBacktrace *previousBacktrace;
 

